chore(core): Move scopes and roles into database in preparation for custom roles

afitzek opened the pull request:
## Summary

We move scopes and roles into the database and use the database as the source of truth for these. This is to
prepare the system for custom roles, which can be stored in the database.

## Related Linear tickets, Github issues, and Community forum posts

closes https://linear.app/n8n/issue/PAY-3054/move-role-definitions-into-the-database

## Review / Merge checklist

- [ ] PR title and summary are descriptive. ([conventions](../blob/master/.github/pull_request_title_conventions.md)) <!--
   **Remember, the title automatically goes into the changelog.
   Use `(no-changelog)` otherwise.**
-->
- [ ] [Docs updated](https://github.com/n8n-io/n8n-docs) or follow-up ticket created.
- [ ] Tests included. <!--
   A bug is not considered fixed, unless a test is added to prevent it from happening again.
   A feature is not complete without tests.
-->
- [ ] PR Labeled with `release/backport` (if the PR is an urgent fix that needs to be backported)

afitzek committed:
Create scope table and sync scopes to db

guillaumejacquart commented on packages/@n8n/permissions/src/scope-information.ts:
```diff
@@ -0,0 +1,24 @@
+
+export const ALL_SCOPES = buildResourceScopes();
+
+export const scopeInformation: Partial<Record<Scope, ScopeInformation>> = {
```
Is that really necessary ?

afitzek replied:
This is just a place holder to describe the scopes for the UI. So I think we should fill this list at some point, when we research all scopes and what they are used for.

guillaumejacquart commented on packages/@n8n/permissions/src/scope-information.ts:
```diff
@@ -0,0 +1,24 @@
+import { RESOURCES } from './constants.ee';
+import type { Scope, ScopeInformation } from './types.ee';
+
+function buildResourceScopes() {
```
What would you think of a more 'declarative' way, like this:
```
function buildResourceScopes(): Scope[] {
	const resourceScopes = Object.entries(RESOURCES).flatMap(
		([resource, operations]) => [
			...operations.map((op) => `${resource}:${op}` as const),
			`${resource}:*` as const,
		],
	) as Scope[];

	resourceScopes.push('*' as const); // Global wildcard
	return resourceScopes;
}
```

Maybe a bit less performant though

guillaumejacquart commented on packages/@n8n/db/src/migrations/common/1750252139168-LinkRoleToUserTable.ts:
```diff
@@ -0,0 +1,78 @@
+					systemRole: true,
+				},
+			);
+		} catch (error) {
```
Agree with the bot here 🤣

guillaumejacquart commented on packages/@n8n/db/src/migrations/common/1750252139168-LinkRoleToUserTable.ts:
```diff
@@ -0,0 +1,79 @@
+		const systemRoleColumn = escape.columnName('systemRole');
+
+		// Make sure that the global roles that we need exist
+		try {
```
DO you think it would make sense to have a loop here (of ['global:owner', ''global:admin', 'global:member']) ?

guillaumejacquart reviewed (commented):
some minor comments

guillaumejacquart reviewed (commented):
Some additional minor comments

afitzek commented on packages/@n8n/db/src/migrations/common/1750252139168-LinkRoleToUserTable.ts:
```diff
@@ -0,0 +1,75 @@
+		await addColumns('user', [column('roleSlug').varchar(128).default("'global:member'").notNull]);
+
+		await runQuery(
+			`UPDATE ${userTableName} SET ${roleSlugColumn} = ${roleColumn} WHERE ${roleColumn} != ${roleSlugColumn}`,
```
In all databases the role column is not nullable. So a null value can't exist in there. 👍

afitzek commented on packages/@n8n/db/src/entities/role.ts:
```diff
@@ -0,0 +1,51 @@
+		default: false,
+		name: 'systemRole',
+	})
+	systemRole: boolean; // Indicates if the role is managed by the system and cannot be edited
```
I think we don't use is in front of bool props. For example finished on the execution entity, or mfaEnabled on the user table.

afitzek commented on packages/@n8n/db/src/migrations/common/1750252139168-LinkRoleToUserTable.ts:
```diff
@@ -0,0 +1,75 @@
+		for (const role of ['global:owner', 'global:admin', 'global:member']) {
+			if (dbType === 'sqlite') {
+				await runQuery(
+					`INSERT OR REPLACE INTO ${tableName} (${slugColumn}, ${roleTypeColumn}, ${systemRoleColumn}) VALUES (:slug, :roleType, :systemRole)`,
```
You are right, I was able to change the insert queries to all ignore on conflict, but still all 3 db queries are different, the sqlite one needs an INSERT OR REPLACE, which postgres doesn't recognize.

guillaumejacquart replied:
I was under the impression that Sqlite did not need INSERT OR REPLACE (see examples here): https://sqlite.org/lang_upsert.html#examples

afitzek replied:
Ah you are right, I miss read 🙈

guillaumejacquart reviewed (dismissed):
looks good to me

guillaumejacquart commented on packages/@n8n/db/src/migrations/common/1750252139168-LinkRoleToUserTable.ts:
```diff
@@ -0,0 +1,66 @@
+		const systemRoleColumn = escape.columnName('systemRole');
+
+		// Make sure that the global roles that we need exist
+		for (const role of ['global:owner', 'global:admin', 'global:member']) {
```
I'm over doing it but what about:
```
const isSqliteOrPostgres = dbType === 'sqlite' || dbType === 'postgresdb';
const query = isSqliteOrPostgres
	? `INSERT INTO ${tableName} (${slugColumn}, ${roleTypeColumn}, ${systemRoleColumn}) VALUES (:slug, :roleType, :systemRole) ON CONFLICT DO NOTHING`
	: `INSERT IGNORE INTO ${tableName} (${slugColumn}, ${roleTypeColumn}, ${systemRoleColumn}) VALUES (:slug, :roleType, :systemRole)`;

for (const role of ['global:owner', 'global:admin', 'global:member']) {
	await runQuery(query, {
		slug: role,
		roleType: 'global',
		systemRole: true,
	});
}
```

tomi commented on packages/cli/src/auth/auth.roles.service.ts:
Very clean and easy to follow code. Nice job 👏

despairblue reviewed (dismissed):
I only looked at the migrations and they look solid. 👍🏾

afitzek commented on packages/@n8n/db/src/migrations/common/1750252139166-AddScopeTables.ts:
I wanted to keep the migrations small and simple and specific to do one thing.

tomi replied:
Personally I'd prefer just one, since every migration is extra overhead in instance startup. One migration can also be composed into smaller functions. But I can understand your reasoning as well.

tomi reviewed (commented):
Thank you for addressing the feedback 💘  LGTM now. Just one question regarding multi-main mode

afitzek commented:
> Thank you for addressing the feedback 💘 LGTM now. Just one question regarding multi-main mode

Thanks, that is a good question:
I think with multi-main it should still be good, because if all mains are running the same version, they try to run the same query, worst case would be that each main runs the same update, but this shouldn't lead to any problems. On the other hand, if only the leader runs this update, with a zero down time update, there can be multiple different versions running at the same time, so we wouldn't be able to ensure that a leader runs this during start up, we would need to run it on each leader change, which would also lead to multiple executions of the query in the end. But if this would become an issue, I think we could look into synchronizing this.

guillaumejacquart commented on packages/@n8n/db/src/migrations/common/1750252139167-AddRolesTables.ts:
```diff
@@ -0,0 +1,98 @@
+
+		await createIndex('role_scope', ['scopeSlug']);
+		/*
+		await createTable('role_scope')
```
Should be remove the comments ?

guillaumejacquart commented on packages/@n8n/db/src/repositories/role.repository.ts:
```diff
@@ -0,0 +1,11 @@
+import { Role } from '../entities';
+
+@Service()
+export class RoleRepository extends Repository<Role> {
```
It will always be time to move later, but I'm just thinking if we should not introduce a "custom role" module or something 🤔

guillaumejacquart reviewed (approved):
Just minor comments. I approve so you can merge when you want

tomi commented:
> > Thank you for addressing the feedback 💘 LGTM now. Just one question regarding multi-main mode
> 
> Thanks, that is a good question: I think with multi-main it should still be good, because if all mains are running the same version, they try to run the same query, worst case would be that each main runs the same update, but this shouldn't lead to any problems. On the other hand, if only the leader runs this update, with a zero down time update, there can be multiple different versions running at the same time, so we wouldn't be able to ensure that a leader runs this during start up, we would need to run it on each leader change, which would also lead to multiple executions of the query in the end. But if this would become an issue, I think we could look into synchronizing this.

It's pretty standard in k8s environments to do a rolling update. And in multi-main mode the workers and mains are different k8s deployments, so they update separately. Meaning there definitely will be cases where one main updates the permissions and other mains & workers are still running the older n8n version.

What's the worst that can happen here? Some permission checks might fail? Some permission checks might succeed when they shouldn't? This is also not a trivial to solve, as it would require versioning the permissions in the DB. I don't really see how synchronization would help here, if we check the permissions against what is in the DB. I guess we can ignore this for now and see if this causes any issues in the future. I hope we don't have to make changes to the permissions that often.
